### PR TITLE
Better error messages for SoundCloud and YouTube unavailable contents

### PR DESCRIPTION
**Pull request by TiA4f8R**
#### What is it?
- [x] Feature (user facing)

#### Description of the changes in your PR
This PR add new error messages (strings added in English and French), on SoundCloud and YouTube services, when a user tries to play:
- a SoundCloud Go+ track
- a SoundCloud track which is blocked in his/her/its country
- a YouTube Music Premium video
- a YouTube paid video (a YouTube movie or a YouTube video which requires a subscription to the video channel)
- a YouTube video which is blocked in his/her/its country (doesn't work for YouTube Music georestricted tracks because of the error message returned by YouTube)
- a YouTube age-restricted video which could not be played, due to YouTube changes for age-restricted videos

#### Screenshots of the error messages

| Screenshot of the error message | Content type of error message   |
|---------------------------------|-----------------------------|
|                                Georestricted content | Georestricted content       |
|                                Private content | Private content             |
|                                Paid content | Paid content                |
|                                YouTube Music Premium video | YouTube Music Premium video |

**TO DO:** update extractor in app/build.gradle when Extractor PR is merged

#### Fixes the following issue(s)
#4979 **for SoundCloud and YouTube**

#### Relies on the following changes
TeamNewPipe/NewPipeExtractor#509

#### APK testing 
[app.zip](https://github.com/TeamNewPipe/NewPipe/suites/2096924712/artifacts/42569341) (from CI)

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Stypox** [comment]
Your approach is the right one. Remember to mark this as Ready when you are finished ;-)

**Comment by TiA4f8R** [comment]
@Stypox The implementation on NewPipe is ready but I need help to parse HLS manifests in the extractor side.

**Comment by Stypox** [comment]
@TiA4f8R is it really a good idea to add support for HLS in a PR about error messages? Also, currently NewPipe's player does not support HLS, so I don't know if your workaround would work.

**Comment by TiA4f8R** [comment]
@Stypox PR is ready for review and workaround for SoundCloud HLS support has been reverted.

File: app/build.gradle [hunk 1]
```diff
@@ -180,7 +180,9 @@ dependencies {
 
     // NewPipe dependencies
     // You can use a local version by uncommenting a few lines in settings.gradle
-    implementation 'com.github.TeamNewPipe:NewPipeExtractor:beb705013c4dbe2c1a893fa338140e7a5176a0d5'
+
+    implementation 'com.github.TiA4f8R:NewPipeExtractor:c20d35d72162819e969157fa240a32149043ccc1'
+
```
Comment (Stypox) [inline app/build.gradle:185]: Do not add blank lines around this

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -699,4 +699,6 @@
     <string name="show_thumbnail_summary">Use thumbnail for both lock screen background and notifications</string>
     <string name="recent">Recent</string>
     <string name="chapters">Chapters</string>
+    <string name="georestricted_content">This content is not available in your country.</string>
+    <string name="soundcloud_go_plus_content">This SoundCloud track is a SoundCloudGo+ track (at least in your country), which cannot be streamed in NewPipe.</string>
```
Comment (Stypox) [inline app/src/main/res/values/strings.xml:703]: (notice the missing point at the end)
```suggestion
    <string name="georestricted_content">This content is not available in your country</string>
    <string name="soundcloud_go_plus_content">This is a SoundCloudGo+ track, at least in your country, so it cannot be streamed by NewPipe</string>
```

**Review by Stypox — CHANGES_REQUESTED** [review]

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -700,4 +700,6 @@
     <string name="recent">Recent</string>
     <string name="chapters">Chapters</string>
     <string name="no_app_to_open_intent">No app on your device can open this</string>
+    <string name="georestricted_content">This content is not available in your country</string>
+    <string name="soundcloud_go_plus_content">This is a SoundCloud track, at least in your country, so it cannot be streamed by NewPipe</string>
```
Comment (Stypox) [inline app/src/main/res/values/strings.xml:704]:
```suggestion
    <string name="soundcloud_go_plus_content">This is a SoundCloudGo+ track, at least in your country, so it cannot be streamed by NewPipe</string>
```

**Review by Stypox — COMMENTED** [review]
I probably suggested the wrong thing before  sorry

**Comment by TiA4f8R** [comment]
No it's me because I keep `Go+` in French

**Comment by Stypox** [comment]
Thank you :-)

**Review by Stypox — DISMISSED** [review]
Thank you :-)

File: app/src/main/java/org/schabi/newpipe/fragments/BaseStateFragment.java [hunk 4]
```diff
@@ -229,6 +231,12 @@ protected boolean onError(final Throwable exception) {
         } else if (ExceptionUtils.isNetworkRelated(exception)) {
             showError(getString(R.string.network_error), true);
             return true;
+        } else if (exception instanceof GeographicRestrictionException) {
+            showError(getString(R.string.georestricted_content), false);
+            return true;
+        } else if (exception instanceof SoundCloudGoPlusException) {
+            showError(getString(R.string.soundcloud_go_plus_content), false);
+            return true;
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/fragments/BaseStateFragment.java:239]: You need to move this up. 
`GeographicRestrictionException` and `SoundCloudGoPlusException` both extend `ContentNotAvailableException`, thus both exceptions go into the first branch.

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -700,4 +700,9 @@
     <string name="recent">Recent</string>
     <string name="chapters">Chapters</string>
     <string name="no_app_to_open_intent">No app on your device can open this</string>
+    <string name="georestricted_content">This content is not available in your country</string>
+    <string name="soundcloud_go_plus_content">This content a SoundCloud Go+ track, at least in your country, so it cannot be streamed or downloaded by NewPipe</string>
```
Comment (opusforlife2) [inline app/src/main/res/values/strings.xml:704]:
```suggestion
    <string name="soundcloud_go_plus_content">This is a SoundCloud Go+ track, at least in your country, so it cannot be streamed or downloaded by NewPipe</string>
```

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -700,4 +700,9 @@
...
+    <string name="private_content">This content is private and cannot be streamed or downloaded by NewPipe</string>
```
Comment (opusforlife2) [inline app/src/main/res/values/strings.xml:705]:
```suggestion
    <string name="private_content">This content is private, so it cannot be streamed or downloaded by NewPipe</string>
```

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -700,4 +700,9 @@
...
+    <string name="youtube_music_premium_content">This video can be only streamed by YouTube Music Premium members, so it cannot be streamed or downloaded by NewPipe</string>
```
Comment (opusforlife2) [inline app/src/main/res/values/strings.xml:706]:
```suggestion
    <string name="youtube_music_premium_content">This video is available only to YouTube Music Premium members, so it cannot be streamed or downloaded by NewPipe</string>
```

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -700,4 +700,9 @@
...
+    <string name="paid_content">This content is paid and cannot be streamed or downloaded by NewPipe</string>
```
Comment (opusforlife2) [inline app/src/main/res/values/strings.xml:707]:
```suggestion
    <string name="paid_content">This content is paid, so it cannot be streamed or downloaded by NewPipe</string>
```

Comment (TobiGr) [reply app/src/main/res/values/strings.xml:707]:
```suggestion
    <string name="paid_content">This content is only available to users who have paid, so it cannot be streamed or downloaded by NewPipe</string>
```

Comment (opusforlife2) [reply app/src/main/res/values/strings.xml:707]: Ah, yes, that's better.

**Comment by TiA4f8R** [comment]
@TobiGr @opusforlife2 I updated the strings based on your suggestions.
The work in NewPipe should be done, now I need to understand why my code doesn't work in the extractor side for videos which requires a paid subscription to the channel which uploaded it.

File: app/src/main/java/org/schabi/newpipe/fragments/BaseStateFragment.java [hunk 9]
```diff
@@ -223,15 +228,30 @@ protected boolean onError(final Throwable exception) {
...
+        } else if (exception instanceof PaidContentException) {
+            showError(getString(R.string.paid_content), false);
+            return true;
         } else if (exception instanceof ContentNotSupportedException) {
             showError(getString(R.string.content_not_supported), false);
             return true;
+        } else if (exception instanceof ContentNotAvailableException) {
+            showError(getString(R.string.content_not_available), false);
+            return true;
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/fragments/BaseStateFragment.java:254]: Please move the `ContentNotAvailableException` branch before the `ContentNotSupportedException`branch to keep the logical order (`ContentNotSupportedException` is not a `ContentNotAvailableException`)

File: app/src/main/java/org/schabi/newpipe/util/ExtractorHelper.java [hunk 10]
```diff
@@ -300,10 +305,22 @@ public static void handleGeneralException(final Context context, final int servi
...
                 Toast.makeText(context, R.string.network_error, Toast.LENGTH_LONG).show();
-            } else if (exception instanceof ContentNotAvailableException) {
-                Toast.makeText(context, R.string.content_not_available, Toast.LENGTH_LONG).show();
+            } else if (exception instanceof SoundCloudGoPlusException) {
+                Toast.makeText(context, R.string.soundcloud_go_plus_content,
+                        Toast.LENGTH_LONG).show();
+            } else if (exception instanceof GeographicRestrictionException) {
+                Toast.makeText(context, R.string.georestricted_content, Toast.LENGTH_LONG).show();
             } else if (exception instanceof ContentNotSupportedException) {
                 Toast.makeText(context, R.string.content_not_supported, Toast.LENGTH_LONG).show();
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/util/ExtractorHelper.java:314]: move this below `ContentNotAvailableException`

**Review by TobiGr — COMMENTED** [review]
Looks good to me. Just a minor thing to improve readability.

**Comment by rocketinventor** [comment]
Hi, I noticed that "restricted" videos (videos with _potentially_  adult content) still don't have an error message (when using YouTube's "restricted mode").

This should be added also to help achieve the goal of helping trouble-shooting and reducing confusion when a video is unavailable.

I added a comment on the original Pull Request, but I thought that this should be addressed here also.

**Review by TobiGr — DISMISSED** [review]

File: app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -164,6 +164,7 @@
     <string name="youtube_restricted_mode_enabled_title">Turn on YouTube\'s \"Restricted Mode\"</string>
     <string name="youtube_restricted_mode_enabled_summary">YouTube provides a \"Restricted Mode\" which hides potentially mature content</string>
     <string name="restricted_video">This video is age restricted.\n\nTurn on \"%1$s\" in the settings if you want to see it.</string>
+    <string name="restricted_video_no_stream">This video is age restricted and no stream could be fetched.</string>
```
Comment (TiA4f8R) [inline app/src/main/res/values/strings.xml:167]: Wouldn't it be better to say something like `Due to new YouTube policies with age-restricted videos, this video cannot be streamed in NewPipe`?

Comment (TobiGr) [reply app/src/main/res/values/strings.xml:167]: Well there are still age-restricted videos which can be played. That's why I wanted to give a reason why this video in particular cannot be played in difference to the other age-restricted videos. Maybe we can combine both error messages:
`This video is age-restricted\nDue to new YouTube policies with age-restricted videos, NewPipe cannot access any of its video streams and thus is unable to play it.`

Comment (TiA4f8R) [reply app/src/main/res/values/strings.xml:167]: I will push a commit

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -700,4 +701,9 @@
     <string name="recent">Recent</string>
     <string name="chapters">Chapters</string>
     <string name="no_app_to_open_intent">No app on your device can open this</string>
+    <string name="georestricted_content">This content is not available in your country</string>
+    <string name="soundcloud_go_plus_content">This is a SoundCloud Go+ track, at least in your country, so it cannot be streamed or downloaded by NewPipe</string>
+    <string name="private_content">This content is private, so it cannot be streamed or downloaded by NewPipe</string>
+    <string name="youtube_music_premium_content">This video is available only to YouTube Music Premium members, so it cannot be streamed or downloaded by NewPipe</string>
+    <string name="paid_content">This content is only available to users who have paid, so it cannot be streamed or downloaded by NewPipe</string>
```
Comment (TobiGr) [inline app/src/main/res/values/strings.xml:708]: These strings need full stops at the end.

**Review by TobiGr — APPROVED** [review]

**Pull request merged**